File: src/core/gui/RendererWidget.ts
import { ResizableWidget } from "./ResizableWidget";
import { div } from "./dom";
import { Widget } from "./Widget";
import { Renderer } from "../rendering/Renderer";

export class RendererWidget extends ResizableWidget {
    readonly element = div({ className: "core_RendererWidget" });
    readonly children: readonly Widget[] = [];

    constructor(private renderer: Renderer) {
        super();

        this.element.append(renderer.canvas_element);

        this.disposable(renderer);

        this.finalize_construction(RendererWidget);
    }

    activate(): void {
        this.renderer.start_rendering();
        super.activate();
    }

    deactivate(): void {
        super.deactivate();
        this.renderer.stop_rendering();
    }

    resize(width: number, height: number): this {
        super.resize(width, height);

        this.renderer.set_size(width, height);

        return this;
    }
}

File: web/src/main/kotlin/world/phantasmal/web/core/widgets/RendererWidget.kt
package world.phantasmal.web.core.widgets

import org.w3c.dom.Node
import world.phantasmal.web.core.rendering.Renderer
import world.phantasmal.webui.dom.div
import world.phantasmal.webui.widgets.Widget

class RendererWidget(
    private val renderer: Renderer,
) : Widget() {
    override fun Node.createElement() =
        div {
            className = "pw-core-renderer"

            observe(selfOrAncestorVisible) { visible ->
                if (visible) {
                    renderer.startRendering()
                } else {
                    renderer.stopRendering()
                }
            }

            addDisposable(size.observe { (size) ->
                renderer.setSize(size.width, size.height)
            })

            append(renderer.canvas)
        }

    companion object {
        init {
            @Suppress("CssUnusedSymbol")
            // language=css
            style("""
                .pw-core-renderer {
                    width: 100%;
                    height: 100%;
                    outline: none;
                }
            """.trimIndent())
        }
    }
}
